hw5_asyncio.py

```diff
@@ -0,0 +1,102 @@
+import aiohttp
+import asyncio
+import os
+import re
+from threading import Thread
+import time
+import urllib.request
+
+
+urls = ['https://www.python.org/ftp/python/3.7.2/Python-3.7.2.tgz',
+        'https://www.python.org/ftp/python/3.6.8/Python-3.6.8.tgz',
+        'https://www.python.org/ftp/python/3.7.2/Python-3.7.2rc1.tgz',
+        'https://www.python.org/ftp/python/3.6.8/Python-3.6.8rc1.tgz',
+        'https://www.python.org/ftp/python/3.7.1/Python-3.7.1.tgz',
+        'https://www.python.org/ftp/python/3.6.7/Python-3.6.7.tgz',
+        'https://www.python.org/ftp/python/3.7.1/Python-3.7.1rc2.tgz',
+        'https://www.python.org/ftp/python/3.6.7/Python-3.6.7rc2.tgz',
+        'https://www.python.org/ftp/python/3.7.1/Python-3.7.1rc1.tgz',
+        'https://www.python.org/ftp/python/3.6.7/Python-3.6.7rc1.tgz']
+
+
+def get_urls(url):
+    request = urllib.request.Request(url)
+    website = urllib.request.urlopen(request)
+    html = str(website.read())
+    links = re.findall(r'"((ht|f)tps?://\S*\.(tgz|tar\.xz))"', html)
+    links = [l[0] for l in links]
+    return links
+
+
+def get_name(link):
+    return link.split('/')[-1]
+
+
+def downloader(link):
+    data = urllib.request.urlopen(link)
+    data_to_write = data.read()
+    with open(str(get_name(link)), 'wb') as f:
+        f.write(data_to_write)
+
+
+def sync_dowload(links):
+    for url in links:
+        downloader(url)
+
+
+def thread_dowload(links):
+    threads = [Thread(target=downloader, args=(link,)) for link in links]
+
+    for th in threads:
```

Comment:
There is no concurrency here as the next thread is not started until the previous one is finished.

```
# run all threads
for th in threads:
        th.start()

# wait until all threads are done
for th in threads:
        th.join()
```